MapleServer2/PacketHandlers/Game/NpcTalkHandler.cs

```diff
@@ -248,7 +209,75 @@ private static void HandleBeauty(GameSession session)
                     break;
             }
             session.Send(UserMoveByPortalPacket.Move(session, portal.Coord.ToFloat(), portal.Rotation.ToFloat()));
+        }
+
+        private static DialogType GetDialogType(ScriptMetadata scriptMetadata, NpcTalk npcTalk, bool hasNextScript)
+        {
+            DialogType dialogType = DialogType.CloseNext1;
+            if (scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId).Goto.Count == 0)
+            {
+                dialogType = DialogType.CloseNext;
+            }
+
+            if (!hasNextScript)
+            {
+                ScriptIdType type = (ScriptIdType) (npcTalk.ScriptId / 100);
+                if (npcTalk.IsQuest)
+                {
+                    dialogType = type switch
+                    {
+                        ScriptIdType.Start => DialogType.AcceptDecline,
+                        ScriptIdType.End => DialogType.QuestReward,
+                        _ => DialogType.Close1,
+                    };
+                }
+                else
+                {
+                    dialogType = DialogType.Close1;
+                }
+            }
+            return dialogType;
+        }
 
+        private static int GetNextScript(ScriptMetadata scriptMetadata, NpcTalk npcTalk, int index)
+        {
+            if (npcTalk.IsQuest && npcTalk.ScriptId == 0)
+            {
+                if (npcTalk.Quests[index].Started)
+                {
+                    // Talking to npc that start the quest and is started
+                    if (npcTalk.Quests[index].StartNpcId == npcTalk.Npc.Id)
+                    {
+                        return 200;
+                    }
+                    else // Talking to npc that end the quest
+                    {
+                        return 300;
+                    }
+                }
+                else // Talking to npc that start the quest and isn't started
+                {
+                    return 100;
+                }
+            }
+            else
+            {
+                if (npcTalk.ScriptId == 0)
+                {
+                    return scriptMetadata.Options.First(x => x.Id > npcTalk.ScriptId).Id;
+                }
+                else
+                {
+                    if (scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId).Goto.Count == 0)
+                    {
+                        return npcTalk.ScriptId;
+                    }
+                    else
+                    {
+                        return scriptMetadata.Options.First(x => x.Id == npcTalk.ScriptId).Goto[index];
```

Comment:
I think you can cache `scriptMetadata.Options.First(x => x.Id > npcTalk.ScriptId)` here too